File: c-kotlin/src/main/kotlin/com/example/HelloController.kt
package com.example

import org.springframework.web.bind.annotation.*
import org.springframework.web.client.RestTemplate

@RestController
class HelloController(private val restTemplate: RestTemplate) {

    @GetMapping("/hello")
    fun sayHello(): String = "Hello from Kotlin microservice!"

    @PostMapping("/handshake")
    fun handshake(@RequestBody text: String): String {
        println("Input from client: $text")

        val request = "$text Hello from Kotlin microservice!"
        val url = "http://host.docker.internal:8083/handshake"
        
        val response = restTemplate.postForObject(url, request, String::class.java) ?: "Error"
        
        println("Output to client: $response")

        return response
    }
}

File: j-scala/src/main/scala/com/example/HelloController.scala
package com.example

import org.springframework.web.bind.annotation._
import org.springframework.web.client.RestTemplate
import org.springframework.beans.factory.annotation.Autowired

@RestController
class HelloController @Autowired()(private val restTemplate: RestTemplate) {

  @GetMapping(Array("/hello"))
  def sayHello(): String = "Hello from Scala microservice!"

  @PostMapping(Array("/handshake"))
  def handshake(@RequestBody text: String): String = {
    println(s"Input from client: $text")
    val request = s"$text Hello from Scala microservice!"
    val url = "http://host.docker.internal:8090/handshake"
    val response = Option(restTemplate.postForObject(url, request, classOf[String])).getOrElse("Error")
    println(s"Output to client: $response")
    response
  }
}
